lib/chromesmith.ex

```diff
@@ -1,30 +1,58 @@
 defmodule Chromesmith do
   @moduledoc """
   Main module for Chromesmith.
+
+  > Add Chromesmith to your application supervision tree:
+
+      defmodule ChromesmithExample.Application do
+        use Application
+
+        def start(_type, _args) do
+          children = [
+            Chromesmith.child_spec(:chrome_pool, [process_pool_size: 2, page_pool_size: 1])
+          ]
+
+          opts = [strategy: :one_for_one, name: ChromesmithExample.Supervisor]
+          Supervisor.start_link(children, opts)
+        end
+      end
+
+  Using `Chromesmith.child_spec/2`, you can provide multiple options to configure the pool:
+
+  * `:process_pool_size` - How many Chrome instances to open
+  * `:page_pool_size` - How many pages to open per chrome instance
+  * `:chrome_options` - Additional chrome headless CLI flags to pass on startup.
   """
   use GenServer
 
   defstruct [
     supervisor: nil, # Chromesmith.Supervisor PID
     process_pool_size: 0, # How many Headless Chrome instances to spawn
-    number_of_pages_per_process: 0, # How many Pages per Instance
-    workers: [], # List of worker tuples, {pid, number_of_used_pages, number_of_total_pages}
-    chrome_options: [] # Options to pass into `:chrome_launcher`
+    page_pool_size: 0, # How many Pages per Instance
+    process_pools: [], # List of process pool tuples, {pid, available_pids, all_pids}
+    chrome_options: [], # Options to pass into `:chrome_launcher`
+    checkout_queue: :queue.new()
   ]
 
   @type t :: %__MODULE__{
     supervisor: pid(),
     process_pool_size: non_neg_integer(),
-    number_of_pages_per_process: non_neg_integer(),
-    workers: [{pid(), non_neg_integer(), non_neg_integer()}]
+    page_pool_size: non_neg_integer(),
+    process_pools: [{pid(), [pid()], [pid()]}],
+    chrome_options: [],
+    checkout_queue: :queue.queue()
   }
 
   @doc """
   Check out a Page from one of the headless chrome processes.
   """
+  @spec checkout(pid(), boolean()) :: {:ok, pid()} | {:error, :none_available} | {:error, :timeout}
   def checkout(pid, should_block \\ true)
-  def checkout(pid, should_block) do
-    GenServer.call(pid, {:checkout, should_block})
+  def checkout(pid, true) do
+    GenServer.call(pid, {:checkout, true}, :infinity)
```

Comment:
this one's a bit scary and could have implications down the road.

The intent of this function is to block until a page is ready, though if a page is never returned, the calling process will be blocked forever.